include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,501 @@
+        set_vec_from_point::apply(point2, p2);
+
+        vec3 base;
+        vec3 v1;
+        vec3 v2;
+        vec3 normal;
+        polygon_2d projected_polygon;
+
+        // Check if the polygon has been projected before
+        if (!boost::geometry::is_empty(m_polygon_projections[polygon_index].projected_polygon))
+        {
+            auto const& projection = m_polygon_projections[polygon_index];
+            base = projection.base;
+            v1 = projection.v1;
+            v2 = projection.v2;
+            normal = projection.normal;
+            projected_polygon = projection.projected_polygon;
+        }
+        else
+        {
+            // TODO: Support interior rings
+            ring_3d const ring = boost::geometry::exterior_ring(polygon);
+            if (boost::size(ring) < detail::minimum_ring_size<ring_3d>::value)
+            {
+                visitor.template apply<failure_few_points_on_face>();
+                return intersection_type::invalid;
+            }
+
+            // Get three non-collinear points from the polygon to define the plane
+            if (!find_non_collinear_points(polygon, base, v1, v2, strategy))
+            {
+                visitor.template apply<failure_collinear_points_on_face>();
+                return intersection_type::invalid;
+            }
+
+            v1 -= base;
+            v2 -= base;
+            normal = cross(v1, v2);
+
+            // Check for non-coplanar points
+            for (auto pit = points_begin(polygon); pit != points_end(polygon); ++pit)
+            {
+                vec3 vi;
+                set_vec_from_point::apply(*pit, vi);
+                auto vol = qvm::dot(normal, vi)- dot(normal, base);
+                if (std::abs(vol) > tolerance)
+                {
+                    visitor.template apply<failure_non_coplanar_points_on_face>();
+                    return intersection_type::invalid;
+                }
+            }
+            // Project the polygon onto the 2D plane defined by v1 and v2
+            for (auto const& point : ring)
+            {
+                vec3 p;
+                set_vec_from_point::apply(point, p);
+                projected_polygon.outer().emplace_back(project_point_to_plane_2d(v1, v2, p));
+            }
+
+            // Check if the projected polygon is valid
+            if (!resolve_dynamic::is_valid<polygon_2d>::apply(projected_polygon, visitor, strategy))
+            {
+                return intersection_type::invalid;
+            }
+
+            // Cache the projection for future use
+            m_polygon_projections[polygon_index] = polygon_projection{projected_polygon, base, v1,
+                    v2, normal};
+        }
+
+        // Check if the segment intersects the plane of the polygon
+        coordinate_type const dot_product = dot(normal, base);
+        coordinate_type const sign1 = dot(normal, p1) - dot_product;
+        coordinate_type const sign2 = dot(normal, p2) - dot_product;
```

Comment:
The `p1` and `p2` are calculated long before...
Would it be possible to move that to here? (And rename them to something vector-like)?